src/ring_theory/dedekind_domain/selmer_group.lean

```diff
@@ -0,0 +1,151 @@
+the `S`-unit group of `R` and `Cl_S(R)` is the `S`-class group of `R`. If the flanking groups are
+finite, then `K(S, n)` is finite by the first isomorphism theorem. Such is the case when `R` is the
+ring of integers of a number field `K`, `S` is finite, and `n` is positive, in which case `R_Sˣ` is
+finitely generated by Dirichlet's unit theorem and `Cl_S(R)` is finite by the class number theorem.
+
+This file defines the Selmer group and some basic facts.
+
+## Main definitions
+
+* `is_dedekind_domain.height_one_spectrum.selmer_group`: the Selmer group.
+* TODO: maps in the sequence
+
+## Main statements
+
+* TODO: proofs of exactness of the sequence
+* TODO: finiteness for global fields
+
+## Notations
+
+* `K⟮S, n⟯`: the Selmer group with parameters `K`, `S`, and `n`.
+
+## Implementation notes
+
+The Selmer group is classically defined in terms of Galois cohomology as a subgroup of `H¹(K, μₙ)`
+under certain local conditions defined by `v`-adic valuations, where `μₙ` is the group of `n`-th
+roots of unity over a separable closure of `K`. Here `H¹(K, μₙ)` is identified with `Kˣ/(Kˣ)ⁿ` by
+the long exact sequence from Kummer theory and Hilbert's theorem 90, and the fundamental exact
+sequence is an easy consequence of the snake lemma. This file will define all the maps explicitly,
+but isomorphisms to the Galois cohomological definition will be provided when possible.
+
+## References
+
+* https://doc.sagemath.org/html/en/reference/number_fields/sage/rings/number_field/selmer_group.html
+
+## Tags
+
+class group, selmer group, unit group
+-/
+
+open_locale classical discrete_valuation non_zero_divisors
+
+universes u v
+
+noncomputable theory
+
+namespace is_dedekind_domain
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R] {K : Type v} [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+local notation K/n := Kˣ ⧸ (pow_monoid_hom n : Kˣ →* Kˣ).range
+
+/-! ### Valuations of non-zero elements -/
+
+namespace height_one_spectrum
+
+/-- The multiplicative `v`-adic valuation on `Kˣ`. -/
+def valuation_of_ne_zero_to_fun (x : Kˣ) : multiplicative ℤ :=
+let hx := is_localization.sec R⁰ (x : K) in multiplicative.of_add $
+  (-(associates.mk v.as_ideal).count (associates.mk $ ideal.span {hx.fst}).factors : ℤ)
+  - (-(associates.mk v.as_ideal).count (associates.mk $ ideal.span {(hx.snd : R)}).factors : ℤ)
+
+lemma valuation_of_ne_zero_to_fun_eq (x : Kˣ) :
+  (v.valuation_of_ne_zero_to_fun x : ℤₘ₀) = v.valuation (x : K) :=
+begin
+  change _ = _ * _,
+  rw [units.coe_inv],
+  change _ = ite _ _ _ * (ite (coe _ = _) _ _)⁻¹,
+  rw [is_localization.to_localization_map_sec,
+      if_neg $ is_localization.sec_fst_ne_zero le_rfl x.ne_zero,
+      if_neg $ non_zero_divisors.coe_ne_zero _],
+  any_goals { exact is_domain.to_nontrivial R },
+  refl
+end
+
+/-- The multiplicative `v`-adic valuation on `Kˣ`. -/
+def valuation_of_ne_zero : Kˣ →* multiplicative ℤ :=
```

Comment:
```suggestion
def valuation_of_units : Kˣ →* multiplicative ℤ :=
```

Reply:
Same here, although in my original file this is defined directly (using the inverse map from `units_with_zero_equiv`, so no need for the separate `fun`), and my main concern is if I should stick with that one instead of this.